statsprinter.go

```diff
@@ -370,141 +317,210 @@ type JSONData struct {
+	hostname, ip string,
+	port uint16,
+	streak uint,
+	rtt float32,
+) {
+	var (
+		// for *bool fields
+		f    = false
+		t    = true
+		data = JSONData{
+			Type:                    probeEvent,
+			Hostname:                hostname,
+			Addr:                    ip,
+			Port:                    port,
+			IsIP:                    &t,
+			Success:                 &t,
+			TotalUnsuccessfulProbes: streak,
+		}
+	)
 
-	ipStr := data.Addr
-	if !j.isIP {
-		data.Hostname = j.hostname
+	if hostname != "" {
 		data.IsIP = &f
-		ipStr = fmt.Sprintf("%s (%s)", data.Hostname, ipStr)
+		data.Message = fmt.Sprintf("Reply from %s (%s) on port %d",
+			hostname, ip, port)
+	} else {
+		data.Message = fmt.Sprintf("Reply from %s on port %d",
+			ip, port)
 	}
 
-	data.Message = fmt.Sprintf("%s from %s on port %d", replyMsg.msg, ipStr, j.port)
+	p.print(data)
+}
 
-	if replyMsg.msg != noReply {
-		data.Latency = replyMsg.rtt
-		data.TotalSuccessfulProbes = j.totalSuccessfulProbes
-		data.Success = &t
+func (p *jsonPrinter) printProbeFail(hostname, ip string, port uint16, streak uint) {
+	var (
+		// for *bool fields
+		f    = false
+		t    = true
+		data = JSONData{
+			Type:                    probeEvent,
+			Hostname:                hostname,
+			Addr:                    ip,
+			Port:                    port,
+			IsIP:                    &t,
+			Success:                 &f,
+			TotalUnsuccessfulProbes: streak,
+		}
+	)
+
+	if hostname != "" {
+		data.IsIP = &f
+		data.Message = fmt.Sprintf("No reply from %s (%s) on port %d",
+			hostname, ip, port)
 	} else {
-		data.TotalUnsuccessfulProbes = j.totalUnsuccessfulProbes
+		data.Message = fmt.Sprintf("No reply from %s on port %d",
+			ip, port)
 	}
 
-	_ = printJson(data)
+	p.print(data)
 }
 
 // printStatistics prints all gathered stats when program exits.
-func (j *statsJsonPrinter) printStatistics() {
+func (p *jsonPrinter) printStatistics(s stats) {
 	data := JSONData{
-		Type:      statsEvent,
-		Message:   fmt.Sprintf("stats for %s", j.hostname),
-		Hostname:  j.hostname,
-		Timestamp: time.Now(),
-
-		StartTimestamp:          &j.startTime,
-		TotalDowntime:           j.totalDowntime.Seconds(),
-		TotalPackets:            j.totalSuccessfulProbes + j.totalUnsuccessfulProbes,
-		TotalSuccessfulProbes:   j.totalSuccessfulProbes,
-		TotalUnsuccessfulProbes: j.totalUnsuccessfulProbes,
-		TotalUptime:             j.totalUptime.Seconds(),
+		Type:     statisticsEvent,
+		Message:  fmt.Sprintf("stats for %s", s.hostname),
+		Hostname: s.hostname,
+
+		StartTimestamp:          &s.startTime,
+		TotalDowntime:           s.totalDowntime.Seconds(),
+		TotalPackets:            s.totalSuccessfulProbes + s.totalUnsuccessfulProbes,
+		TotalSuccessfulProbes:   s.totalSuccessfulProbes,
+		TotalUnsuccessfulProbes: s.totalUnsuccessfulProbes,
+		TotalUptime:             s.totalUptime.Seconds(),
 	}
 
 	data.TotalPacketLoss = (float32(data.TotalUnsuccessfulProbes) / float32(data.TotalPackets)) * 100
```

Comment:
while running some tests, I noticed this is not similar to the planePrinter in which it prints more than 3 fractional parts.

Reply:
Will check

Reply:
Only now I realized that the way I decided to handle numbers in JSON is probably completely wrong.

![image](https://github.com/pouriyajamshidi/tcping/assets/5007271/1b2b0a53-abb6-4ab2-9d69-4f15a25e6d9c)

The initial issue is fixed, but you notice that some numbers (the ones we truncated to 3 fractional points) are represented as strings, which is not good for log-scrappers.